python-api/src/main/python/livy/client.py

```diff
@@ -374,71 +362,79 @@ def _load_config(self, load_defaults=True, conf_dict=None):
         job_handle = JobHandle(self._conn, self._session_id, self._executor)
-        job_handle._start(command, base64_pickled_job_json)
+        job_handle._start(command, base64_pickled_job_data)
         return job_handle
 
     def _add_file_or_pyfile_job(self, command, file_uri):
-        json_data = json.dumps({'uri': file_uri})
+        data = {'uri': file_uri}
         suffix_url = "/" + str(self._session_id) + "/" + command
-        header = {'Content-Type': 'application/json', 'X-Requested-By': 'livy'}
-        return self._executor.submit(self._add_or_upload_resource, suffix_url,
-                                     data=json_data, headers=header)
+        return self._executor.submit(
+            self._add_or_upload_resource, suffix_url, data=data,
+            headers=self._conn._JSON_HEADERS)
 
     def _upload_file_or_pyfile(self, command, open_file):
         files = {'file': open_file}
         suffix_url = "/" + str(self._session_id) + "/" + command
-        headers = {'X-Requested-By': 'livy'}
-        return self._executor.submit(self._add_or_upload_resource, suffix_url,
-                                     files=files, headers=headers)
+        return self._executor.submit(
+            self._add_or_upload_resource, suffix_url, files=files)
 
-    def _add_or_upload_resource(self, suffix_url, **kwargs):
-        return self._conn.send_request('POST', suffix_url, **kwargs).content
+    def _add_or_upload_resource(self, suffix_url, files=None, data=None,
+                                headers={}):
+        return self._conn.send_request(
+            'POST', suffix_url, files=files, data=data,
+            headers=headers).content
 
 
-class LivyConnection(object):
+class _LivyConnection(object):
 
     _SESSIONS_URI = '/sessions'
     # Timeout in seconds
     _TIMEOUT = 10
+    _JSON_HEADERS = {'Content-Type': 'application/json',
+                     'Accept': 'application/json'}
 
     def __init__(self, uri):
         self._server_url_prefix = uri.geturl() + self._SESSIONS_URI
         self._requests = requests
         self.lock = threading.Lock()
 
-    def send_request(self, method, suffix_url, **kwargs):
-        with self.lock:
-            request_url = self._server_url_prefix + suffix_url
-            response = self._requests.request(method, request_url,
-                                              timeout=self._TIMEOUT, **kwargs)
-            return response
+    def send_request(
+            self, method, suffix_url, headers={}, files=None, data=None):
+        try:
+            with self.lock:
+                headers['X-Requested-By'] = 'livy'
+                request_url = self._server_url_prefix + suffix_url
+                return self._requests.request(
+                    method, request_url, timeout=self._TIMEOUT,
+                    headers=headers, files=files, json=data)
+        finally:
+            if files is not None:
+                files.clear()
```

Comment:
files is an input parameter, so this method does not own it. meaning it shouldn't be changed, unless you explain the behavior in the method's documentation.